### DISCUSSION TITLE
Problem with time-consuming operations on OM protocol side

### DISCUSSION THREAD

**Discussion by vtutrinov** [apache/ozone#7357] [FAQ] [answered]
## Preamble
There was quite a large load for the OM Leader through writing keys by S3G MultipartUpload operations: 14  100Gb files were written to the cluster from 14 clients/hosts (`aws s3 cp`)

## Performance problems
During the writing operations, OM Leader's performance was awkward:
* The AllocateBlock operation could take up to ~30 sec.
![380067163-651baae4-84e8-45d9-82b8-8ef9bd1fb186](https://github.com/user-attachments/assets/d1e58b80-bea1-4e29-be93-5f82cf7285d5)

* Almost 99% of the time was consumed by OM protocol messages handling
![2024-10-25-110901_1582x176_scrot](https://github.com/user-attachments/assets/fccaac21-decc-4482-a5d5-68ba4ef11c0d)
![2024-10-25-110920_3683x1339_scrot](https://github.com/user-attachments/assets/597fe4ac-34d4-4567-80ab-88e12f256a56)
![2024-10-25-093850_3692x1563_scrot](https://github.com/user-attachments/assets/798094cf-4d5b-4b76-9ae6-4e0557eaa4b2)
![2024-10-25-093908_3702x1546_scrot](https://github.com/user-attachments/assets/a972fee0-c738-491f-a616-4f12ca73c3c5)

Does anybody know how to try to work the issue around?
I would suggest the following steps:
* try to reduce message size (e.g. cut out optional fields)
* muti-raft? At the moment only one OM instance handles the transactions (the other two instances do nothing)

**Comment by ivandika3** [comment, marked as answer]
There are some existing tickets to optimize the space for Multipart Upload including things like reducing the message size:
- https://issues.apache.org/jira/browse/HDDS-8238
- https://issues.apache.org/jira/browse/HDDS-10611

As I understand it, the main issue is that currently `multipartInfoTable` stores `OmMultipartKeyInfo` which contains all the previously committed parts for a particular upload ID. This means that as more parts are committed, the serde cost would increase considerably as we need to load and write the existing parts. As suggested by HDDS-8238, we can change the structure of the multipart info table (we probably need to introduce a new table) to "flatten" the structure such that each DB entry only contains a single part (e.g. maybe with the DB key "{keyName}/{uploadId}/{partNumber}" so that we can reduce the serde cost. However, this requires some significant change from OM and possibly client side. 

As you suggested, removing unnecessary fields might also help to reduce the serde cost. As seen from the profile `OzoneAclInfo` might be removed since it's already in the `OpenKeyTable`. Maybe we can just set the `PartKeyInfo` ACL to empty list before committing it.

As for the multi-raft solution, this probably requires more engineering effort since this means sharding the OMs which requires more deep design decisions. Even if for example we shard by bucket, if there are many concurrent multipart uploads in the bucket, the issue would happen again.

@vtutrinov As for now, do we know how large is the multipart uploads part size? Smaller part size might cause increase in the number of MPU part generated which also adds the serde cost, so maybe you can increase the size of the parts for large MPU key. 

For the long AllocateBlock calls, I'm not really sure since part upload will specify the size and will use CreateKey instead. Unless your 
part size is very large and exceed preallocated block max `ozone.key.preallocation.max.blocks`, which I doubt will happen. Maybe there are some performance issues for SCM block allocation process or some datanodes issue that require additional blocks from AllocateBlocks.

Feel free to pick the tickets or file your own if needed.

**Reply by errose28** [reply]
+1. This is similar to the previous discussion [here](https://github.com/apache/ozone/discussions/6869#discussioncomment-9898530). The MPU code could definitely use some streamlining to better handle smaller part sizes, which are the default with most S3 clients.

**Reply by vtutrinov** [reply]
@ivandika3 @errose28 As OzoneAclInfo and DatanodeDetailProto serialization consumes most of the time, and the pipeline field of the OmKeyLocationInfo is an optional one (the DatanodeDetailsProto message is a part of the pipeline), we've tried to set these field to null:

```
      //Set the data size and location info list
      omKeyInfo.setDataSize(keyArgs.getDataSize());
      omKeyInfo.setAcls(null);
      omKeyInfo.updateLocationInfoList(keyArgs.getKeyLocationsList().stream()
          .map(keyLocation -> {
            OmKeyLocationInfo omKeyLocationInfo = OmKeyLocationInfo.getFromProtobuf(keyLocation);
            omKeyLocationInfo.setPipeline(null); // attempt to reduce response size
            return omKeyLocationInfo;
          }).collect(Collectors.toList()), true);
``` 

and the speed of the file uploading (14 clients from 14 hosts tried to upload 100G files) had sped up to 10x.
### Question:
Are there any cons or ineligible effects of such changes? We know that the ACLs are in another RDB table, what about the pipeline? (FYI: tests of writing, reading, and deleting the keys committed with the changes above were passed successfully)

**Comment by vtutrinov** [comment]
@ivandika3, thanks for the quick response! A workaround involving increasing the part size during MPU works well. It seems the fix  for the HDDS-8238 jira ticket would be a great way to resolve the problem

**Comment by spacemonkd** [comment]
Hi @vtutrinov, we have started the work on fixing:
- https://issues.apache.org/jira/browse/HDDS-8238
- https://issues.apache.org/jira/browse/HDDS-10611

Hopefully the next release should have the fixes in place.
I am closing this discussion as it seems your question has been answered. Please feel free to reopen in case you have any other questions.
